server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -658,68 +673,257 @@ protected void doRun() {
+            }
+        }
+
+        private void performInferenceOnBulkItemRequest(
+            BulkShardRequest bulkShardRequest,
+            BulkItemRequest request,
+            Map<String, Set<String>> fieldsForModels,
+            Releasable releaseOnFinish
+        ) {
+            DocWriteRequest<?> docWriteRequest = request.request();
+            Map<String, Object> sourceMap = null;
+            if (docWriteRequest instanceof IndexRequest indexRequest) {
+                sourceMap = indexRequest.sourceAsMap();
+            } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+                sourceMap = updateRequest.docAsUpsert() ? updateRequest.upsertRequest().sourceAsMap() : updateRequest.doc().sourceAsMap();
+            }
+            if (sourceMap == null || sourceMap.isEmpty()) {
+                releaseOnFinish.close();
+                return;
+            }
+            final Map<String, Object> docMap = new ConcurrentHashMap<>(sourceMap);
+
+            // When a document completes processing, update the source with the inference
+            try (var docRef = new RefCountingRunnable(() -> {
+                if (docWriteRequest instanceof IndexRequest indexRequest) {
+                    indexRequest.source(docMap);
+                } else if (docWriteRequest instanceof UpdateRequest updateRequest) {
+                    if (updateRequest.docAsUpsert()) {
+                        updateRequest.upsertRequest().source(docMap);
+                    } else {
+                        updateRequest.doc().source(docMap);
+                    }
+                }
+                releaseOnFinish.close();
+            })) {
+
+                for (Map.Entry<String, Set<String>> fieldModelsEntrySet : fieldsForModels.entrySet()) {
+                    String modelId = fieldModelsEntrySet.getKey();
+
+                    @SuppressWarnings("unchecked")
+                    Map<String, Object> rootInferenceFieldMap = (Map<String, Object>) docMap.computeIfAbsent(
+                        ROOT_RESULT_FIELD,
+                        k -> new HashMap<String, Object>()
+                    );
+
+                    List<String> inferenceFieldNames = getFieldNamesForInference(fieldModelsEntrySet, docMap);
+
+                    if (inferenceFieldNames.isEmpty()) {
+                        continue;
+                    }
+
+                    docRef.acquire();
+
+                    InferenceAction.Request inferenceRequest = new InferenceAction.Request(
+                        TaskType.SPARSE_EMBEDDING, // TODO Change when task type doesn't need to be specified
+                        modelId,
+                        inferenceFieldNames.stream().map(docMap::get).map(String::valueOf).collect(Collectors.toList()),
+                        Map.of()
+                    );
+
+                    client.execute(InferenceAction.INSTANCE, inferenceRequest, new ActionListener<>() {
```

Comment:
Note that I opened https://github.com/elastic/elasticsearch/issues/103665 to discuss how we should handle the additional for bulk/ingest inference execution. I think we need a better strategy to not overload the inference API with a single bulk request. I am investigating this aspect in parallel, no need to do anything in this PR yet ;)